### PR TITLE
Update NGINX and modules

### PR DESCRIPTION
**Pull request by felixbuenemann**
* Update to NGINX 1.14.2 – notable changes since 1.13.7:
  - [HTTP/2 server push support](https://www.nginx.com/blog/nginx-1-13-9-http2-server-push/)  (we should add an annotation to enable `http2_push_preload`)
  - Support for the binary  PROXY protocol v2 in the `proxy_protocol` directive (this is required for using PROXY protocol with AWS Network Load Balancer. Classic ELB uses the plaintext v1 protocol.)
  - Fixes [CVE-2018-16843](https://nvd.nist.gov/vuln/detail/CVE-2018-16843) and [CVE-2018-16844](https://nvd.nist.gov/vuln/detail/CVE-2018-16844) DoS attacks via HTTP/2
* Update VTS module to 0.1.18 (lots of [crash-fixes](https://github.com/vozlt/nginx-module-vts/blob/master/Changes))
* Update to last go-dev image based on deis/base image (go 1.7 -> 1.10)
* Don't bake Dockerfile into image
* Remove external binaries, use multi-stage build
  - libmodsecurity 3.0.3 is built in a separate stage
  - remove proprietary tcell nginx ext
  - remove precompiled modsecurity binaries
  - add geoip2 module (because geoip is deprecated)
  - add gunzip core module (useful if proxying precompressed assets)
  - update OWASP mod security core ruleset to latest 3.2/dev
  - remove external configs from s3, copy or modify templates
  - fix some uncaught errors in build script that could lead
    to corrupted images, because a failure would not set
    an error exit code
  - log all shell commands via `set -x`
  - speed up make by using parallel jobs based on cpu cores (`nproc`)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,7 +12,7 @@ REPO_PATH := github.com/deis/${SHORT_NAME}
 
 # The following variables describe the containerized development environment
 # and other build options
-DEV_ENV_IMAGE := quay.io/deis/go-dev:v0.22.0
+DEV_ENV_IMAGE := quay.io/deis/go-dev:v1.10.0
```
Comment (Cryptophobia) [inline Makefile]: Very nice update of the dev_env image. :+1:

File: model/model.go [hunk 2]
```diff
@@ -436,7 +436,7 @@ func linkLocations(appConfigs []*AppConfig) error {
 		if app.ProxyDomain != "" && len(app.ProxyLocations) > 0 {
 			targetApp := appByDomain(appConfigs, app.ProxyDomain)
 			if targetApp == nil {
-				return goerrors.New(fmt.Sprintf("Can't find ProxyDomain '%s' in any application", app.ProxyDomain))
+				return fmt.Errorf("Can't find ProxyDomain '%s' in any application", app.ProxyDomain)
```
Comment (Cryptophobia) [inline model/model.go]: Thanks for cleaning this up here to proper go error formatting. :)

File: rootfs/Dockerfile [hunk 3]
```diff
@@ -7,20 +41,33 @@ RUN adduser --system \
...
+    runtimeDeps='libcurl3 libxml2 libpcre3 libgeoip1 libmaxminddb0 libfuzzy2 openssl' && \
     apt-get update && \
     apt-get install -y --no-install-recommends \
         $buildDeps \
-        libgeoip1 && \
-    export NGINX_VERSION=1.13.7 SIGNING_KEY=A1C052F8 VTS_VERSION=0.1.10 BUILD_PATH=/tmp/build PREFIX=/opt/router && \
+        $runtimeDeps && \
+    export NGINX_VERSION=1.14.2 SIGNING_KEY=A1C052F8 \
+           VTS_VERSION=0.1.18 GEOIP2_VERSION=3.2 \
+           MOD_SECURITY_NGINX_VERSION=d7101e13685efd7e7c9f808871b202656a969f4b \
```
Comment (Cryptophobia) [inline rootfs/Dockerfile]: I like this a lot. Now we can clearly see what tag of the modsecurity code we are pulling. :+1:

**Comment by Cryptophobia** [comment]
Great work on all of the updates, modsecurity cleanup, multi-stage build. Also, thanks for the organized bullet point write up. :1st_place_medal: 

I'm going to test this a bit today in AWS and minikube and then should be merged in after that.

File: Makefile (hunk 1 above)
Comment (felixbuenemann) [reply Makefile]: Yeah, this is a conservative change. We could probably use the newest version, but it uses `ubuntu:xenial` instead `deis/base:v0.3.6` (also a stripped down xenial).

Comment (Cryptophobia) [reply Makefile]: Let's start conservatively for now and then we can move up. :+1:

**Comment by Cryptophobia** [comment]
Alright, did some tests on minikube, AWS, and GKE. Didn't test edge cases but router is stable with a simple configuration and ModSecurity still works well. :+1: 

Merged!
